File: src/pricing.ts
/**
 * Token pricing for usage accounting.
 *
 * The broker computes cost itself rather than trusting whatever an agent reports,
 * so a misbehaving or simply sloppy harness cannot skew the ledger. Rates are list
 * prices in USD per million tokens and are *defaults only* — anything in
 * `models.<id>.pricing` in agent-bus.config.json wins, which is where you correct
 * a rate that has moved or add a model this table has never heard of.
 *
 * Subscription-plan models (Claude Code Max, a Codex plan) cost nothing per call.
 * For those the ledger still computes what the same tokens would have cost on the
 * metered API and files it as `notionalUSD`, so a swarm running on a flat-rate plan
 * still reports what it consumed. Only metered traffic lands in `costUSD`.
 */
import type { BusConfig, ModelDefinition } from "./config.js";

export interface TokenPricing {
  /** USD per million uncached input tokens. */
  inputPerMTok: number;
  /** USD per million output tokens (reasoning tokens bill as output). */
  outputPerMTok: number;
  /** USD per million tokens read from cache. Defaults to 10% of input. */
  cacheReadPerMTok?: number;
  /** USD per million tokens written to cache. Defaults to 125% of input. */
  cacheWritePerMTok?: number;
}

export interface TokenCounts {
  inputTokens: number;
  outputTokens: number;
  cacheReadTokens: number;
  cacheWriteTokens: number;
  reasoningTokens: number;
}

export interface CostBreakdown {
  /** Billed to a metered API key. Zero for subscription-plan models. */
  costUSD: number;
  /** What these tokens would cost at metered rates, whatever the plan. */
  notionalUSD: number;
  /** Which pricing row produced the number, for auditing a surprising total. */
  pricingSource: "config" | "table" | "unknown";
  billing: "metered" | "subscription" | "local";
}

/**
 * Substring-matched so `claude-opus-4-8`, `anthropic/claude-opus-4.8` and
 * `claude-opus-4-8-20260101` all resolve to the same row. Longest match wins, so
 * a specific entry always beats the family fallback above it.
 */
const TABLE: Array<[string, TokenPricing]> = [
  // Anthropic
  ["claude-opus", { inputPerMTok: 15, outputPerMTok: 75, cacheReadPerMTok: 1.5, cacheWritePerMTok: 18.75 }],
  ["claude-sonnet", { inputPerMTok: 3, outputPerMTok: 15, cacheReadPerMTok: 0.3, cacheWritePerMTok: 3.75 }],
  ["claude-haiku", { inputPerMTok: 1, outputPerMTok: 5, cacheReadPerMTok: 0.1, cacheWritePerMTok: 1.25 }],
  ["claude-3-5-haiku", { inputPerMTok: 0.8, outputPerMTok: 4, cacheReadPerMTok: 0.08, cacheWritePerMTok: 1 }],
  // OpenAI
  ["gpt-5-codex", { inputPerMTok: 1.25, outputPerMTok: 10, cacheReadPerMTok: 0.125 }],
  ["gpt-5-mini", { inputPerMTok: 0.25, outputPerMTok: 2, cacheReadPerMTok: 0.025 }],
  ["gpt-5-nano", { inputPerMTok: 0.05, outputPerMTok: 0.4, cacheReadPerMTok: 0.005 }],
  ["gpt-5", { inputPerMTok: 1.25, outputPerMTok: 10, cacheReadPerMTok: 0.125 }],
  ["gpt-4.1-mini", { inputPerMTok: 0.4, outputPerMTok: 1.6, cacheReadPerMTok: 0.1 }],
  ["gpt-4.1", { inputPerMTok: 2, outputPerMTok: 8, cacheReadPerMTok: 0.5 }],
  ["o3", { inputPerMTok: 2, outputPerMTok: 8, cacheReadPerMTok: 0.5 }],
  // Google
  ["gemini-2.5-pro", { inputPerMTok: 1.25, outputPerMTok: 10, cacheReadPerMTok: 0.31 }],
  ["gemini-2.5-flash", { inputPerMTok: 0.3, outputPerMTok: 2.5, cacheReadPerMTok: 0.075 }],
  ["gemini-3", { inputPerMTok: 2, outputPerMTok: 12, cacheReadPerMTok: 0.5 }],
  // Open-weight hosts
  ["deepseek", { inputPerMTok: 0.28, outputPerMTok: 0.42, cacheReadPerMTok: 0.028 }],
  ["qwen", { inputPerMTok: 0.4, outputPerMTok: 1.2 }],
  ["kimi", { inputPerMTok: 0.6, outputPerMTok: 2.5 }],
  ["glm", { inputPerMTok: 0.6, outputPerMTok: 2.2 }],
  ["llama", { inputPerMTok: 0.2, outputPerMTok: 0.6 }],
];

/** Models that run on the user's own hardware never cost anything. */
const LOCAL_PROVIDERS = new Set(["ollama", "lmstudio", "local", "llamacpp", "vllm"]);

export function lookupPricing(modelName: string): { pricing: TokenPricing; source: "table" | "unknown" } {
  const needle = modelName.toLowerCase();
  let best: TokenPricing | null = null;
  let bestLength = 0;
  for (const [key, pricing] of TABLE) {
    if (needle.includes(key) && key.length > bestLength) {
      best = pricing;
      bestLength = key.length;
    }
  }
  if (best) return { pricing: best, source: "table" };
  // Unknown model: bill it as nothing rather than inventing a rate. The event still
  // records the tokens, so the gap is visible as tokens-without-cost in the summary.
  return { pricing: { inputPerMTok: 0, outputPerMTok: 0 }, source: "unknown" };
}

function rate(pricing: TokenPricing, counts: TokenCounts): number {
  const cacheRead = pricing.cacheReadPerMTok ?? pricing.inputPerMTok * 0.1;
  const cacheWrite = pricing.cacheWritePerMTok ?? pricing.inputPerMTok * 1.25;
  const millions = (n: number) => Math.max(0, n) / 1_000_000;
  return (
    millions(counts.inputTokens) * pricing.inputPerMTok +
    millions(counts.outputTokens + counts.reasoningTokens) * pricing.outputPerMTok +
    millions(counts.cacheReadTokens) * cacheRead +
    millions(counts.cacheWriteTokens) * cacheWrite
  );
}

export function costFor(model: ModelDefinition | null, modelName: string, counts: TokenCounts): CostBreakdown {
  const configured = model?.pricing;
  const { pricing, source } = configured
    ? { pricing: configured, source: "config" as const }
    : lookupPricing(model?.exactModel ?? modelName);

  const provider = (model?.provider ?? "").toLowerCase();
  const billing: CostBreakdown["billing"] = LOCAL_PROVIDERS.has(provider)
    ? "local"
    : model?.capabilities?.costClass === "subscription"
      ? "subscription"
      : "metered";

  const notionalUSD = billing === "local" ? 0 : round6(rate(pricing, counts));
  return {
    costUSD: billing === "metered" ? notionalUSD : 0,
    notionalUSD,
    pricingSource: source,
    billing,
  };
}

/** Resolve a model id or an exact model name against the configured catalogue. */
export function findModel(config: BusConfig, modelName: string): ModelDefinition | null {
  if (!modelName) return null;
  const direct = config.models?.[modelName];
  if (direct) return direct;
  const needle = modelName.toLowerCase();
  for (const model of Object.values(config.models ?? {})) {
    if (model.exactModel && model.exactModel.toLowerCase() === needle) return model;
  }
  return null;
}

function round6(value: number): number {
  return Math.round(value * 1e6) / 1e6;
}

File: doohickey/Sources/Core/Pricing.swift
import Foundation

/// List-price table, USD per million tokens.
///
/// Deliberately the same numbers as the broker's `src/pricing.ts`, so a swarm run
/// costed by agent-bus and the same run costed from Claude Code transcripts agree.
/// Overridable at `~/Library/Application Support/Doohickey/pricing.json` for when a
/// rate moves or a model lands that this table has never heard of.
enum Pricing {
    struct Rate: Codable, Sendable {
        var input: Double
        var output: Double
        var cacheRead: Double?
        var cacheWrite: Double?

        func cost(_ tokens: TokenCounts) -> Double {
            let read = cacheRead ?? input * 0.1
            let write = cacheWrite ?? input * 1.25
            let m = { (n: Int) in Double(max(0, n)) / 1_000_000 }
            return m(tokens.input) * input
                + m(tokens.output + tokens.reasoning) * output
                + m(tokens.cacheRead) * read
                + m(tokens.cacheWrite) * write
        }
    }

    /// Longest substring match wins, so `gpt-5-mini` never gets priced as `gpt-5`.
    private static let builtin: [(String, Rate)] = [
        ("claude-opus", Rate(input: 15, output: 75, cacheRead: 1.5, cacheWrite: 18.75)),
        ("claude-sonnet", Rate(input: 3, output: 15, cacheRead: 0.3, cacheWrite: 3.75)),
        ("claude-haiku", Rate(input: 1, output: 5, cacheRead: 0.1, cacheWrite: 1.25)),
        ("claude-3-5-haiku", Rate(input: 0.8, output: 4, cacheRead: 0.08, cacheWrite: 1)),
        ("gpt-5-codex", Rate(input: 1.25, output: 10, cacheRead: 0.125, cacheWrite: nil)),
        ("gpt-5-mini", Rate(input: 0.25, output: 2, cacheRead: 0.025, cacheWrite: nil)),
        ("gpt-5-nano", Rate(input: 0.05, output: 0.4, cacheRead: 0.005, cacheWrite: nil)),
        ("gpt-5", Rate(input: 1.25, output: 10, cacheRead: 0.125, cacheWrite: nil)),
        ("gpt-4.1-mini", Rate(input: 0.4, output: 1.6, cacheRead: 0.1, cacheWrite: nil)),
        ("gpt-4.1", Rate(input: 2, output: 8, cacheRead: 0.5, cacheWrite: nil)),
        ("o3", Rate(input: 2, output: 8, cacheRead: 0.5, cacheWrite: nil)),
        ("gemini-2.5-pro", Rate(input: 1.25, output: 10, cacheRead: 0.31, cacheWrite: nil)),
        ("gemini-2.5-flash", Rate(input: 0.3, output: 2.5, cacheRead: 0.075, cacheWrite: nil)),
        ("gemini-3", Rate(input: 2, output: 12, cacheRead: 0.5, cacheWrite: nil)),
        ("deepseek", Rate(input: 0.28, output: 0.42, cacheRead: 0.028, cacheWrite: nil)),
        ("qwen", Rate(input: 0.4, output: 1.2, cacheRead: nil, cacheWrite: nil)),
        ("kimi", Rate(input: 0.6, output: 2.5, cacheRead: nil, cacheWrite: nil)),
        ("glm", Rate(input: 0.6, output: 2.2, cacheRead: nil, cacheWrite: nil)),
        ("llama", Rate(input: 0.2, output: 0.6, cacheRead: nil, cacheWrite: nil)),
    ]

    private static let overrides: [String: Rate] = {
        let url = FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Doohickey/pricing.json")
        guard let data = try? Data(contentsOf: url),
              let decoded = try? JSONDecoder().decode([String: Rate].self, from: data)
        else { return [:] }
        return decoded
    }()

    private static var memo: [String: Rate?] = [:]
    private static let lock = NSLock()

    static func rate(for model: String) -> Rate? {
        let needle = model.lowercased()
        lock.lock()
        defer { lock.unlock() }
        if let cached = memo[needle] { return cached }

        var best: Rate?
        var bestLength = 0
        for (key, rate) in overrides where needle.contains(key.lowercased()) && key.count > bestLength {
            best = rate
            bestLength = key.count
        }
        if best == nil {
            for (key, rate) in builtin where needle.contains(key) && key.count > bestLength {
                best = rate
                bestLength = key.count
            }
        }
        memo[needle] = best
        return best
    }

    /// Zero for a model with no known rate — an invented price is worse than a gap,
    /// and the token count is still shown either way.
    static func cost(model: String, tokens: TokenCounts) -> Double {
        rate(for: model)?.cost(tokens) ?? 0
    }
}
